src/swarm/neo/util/VoidBufferAsArrayOf.d

```diff
@@ -0,0 +1,210 @@
+        auto s_array = cast(S[])*void_array;
+    ---
+
+    The helper VoidBufferAsArrayOf simplifies this procedure and removes the
+    risk of undefined behaviour by always handling the void[] as a void[]
+    internally.
+
+    Copyright: Copyright (c) 2017 sociomantic labs GmbH. All rights reserved
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module swarm.neo.util.VoidBufferAsArrayOf;
+
+import ocean.transition;
+
+/*******************************************************************************
+
+    Struct template which wraps a void[] with an API allowing it to be safely
+    used as an array of another type.
+
+    Params:
+        T = element type of array which the API mimics
+
+*******************************************************************************/
+
+public struct VoidBufferAsArrayOf ( T )
+{
+    // T == void is not only pointless but is also invalid: it's illegal to pass
+    // a void argument to a function (e.g. opCatAssign).
+    static assert(!is(T == void));
+
+    /// Pointer to the unerlying void buffer. Must be set before use by struct
+    /// construction.
+    private void[]* buffer;
+
+    // The length of the buffer must always be an even multiple of T.sizeof.
+    invariant ( )
+    {
+        assert(this.buffer.length % T.sizeof == 0);
+    }
+
+    /***************************************************************************
+
+        Returns:
+            a slice of this array
+
+    ***************************************************************************/
+
+    public T[] array ( )
+    {
+        return cast(T[])(*this.buffer);
+    }
+
+    /***************************************************************************
+
+        Returns:
+            the number of T elements in the array
+
+    ***************************************************************************/
+
+    public size_t length ( )
+    {
+        return this.buffer.length / T.sizeof;
+    }
+
+    /***************************************************************************
+
+        Sets the length of the array.
+
+        Params:
+            len = new length of the array, in terms of the number of T elements
+
+    ***************************************************************************/
+
+    public void length ( size_t len )
+    {
+        this.buffer.length = len * T.sizeof;
+        enableStomping(*this.buffer);
+    }
+
+    /***************************************************************************
+
+        Appends an array of elements.
+
+        Params:
+            arr = elements to append
+
+        Returns:
+            a slice of this array, now with the specified elements appended
+
+    ***************************************************************************/
+
+    public T[] opCatAssign ( in T[] arr )
+    {
+        return cast(T[])((*this.buffer) ~= arr);
```

Comment:
This needs `cast(void[])arr` i.e. casting `const` away from `arr` and its elements.

To avoid undefined behaviour (in particular, if `T` contains `const` references, modifying the referenced objects) the elements of `(*this.buffer)` may only be accessed through `cast(T[])(*this.buffer)`, which restores the correct `const`ness of `T`. Making sure of that is the whole point of this struct, but maybe state this in the documentation of the `buffer` field.

Reply:
The constness of `T` is a different question to the consteness of this method's argument, or?

Reply:
Yes, but the constness of `T` is what matters most. We only read `arr` to store a copy of it, but `in` makes the elements of `arr` always `const`, no matter if `T` is `const` or not.

Reply:
Updated.